### ISSUE TITLE
Workflow for cross-compiling, development

### ISSUE THREAD

**Issue by mlund** [dirkwhoffmann/virtualc64#362]
Similarly to closed issue #256, I have a setup for cross-compiling 6510 code in MacOS and I'm after a workflow to test an executable on a freshly restarted C64. In v2.2,

./VirtualC64 ~/code/MARKS.prg
...
D64Archive: Creating D64 archive from a PRG archive...
Disk: Encoded 61488 bits (7686 bytes) for tracks 1 to 17.
Disk: Encoded 57000 bits (7125 bytes) for track 18 (directory).
Disk: Encoded 57000 bits (7125 bytes) for tracks 19 to 24.
Disk: Encoded 53280 bits (6660 bytes) for tracks 25 to 30.
Disk: Encoded 49912 bits (6239 bytes) for tracks 31 and above.

converts the prg file to an attached disk image and I would then do load "*",8,1. Ideally I'd like to flush the prg file directly to memory, possibly from the command line, but a keyboard shortcut in an already open VirtualC64 session would also be a good option as there's little need to restart the GUI in between runs.

PS: Great project!

**Comment by dirkwhoffmann** [comment]
In the next release, PRG and P00 files will be flushed into memory immediately when mount dialogs are disabled in the emulator preferences.

**Comment by mlund** [comment]
Many thanks, I'm looking forward to v2.3.

**Comment by dirkwhoffmann** [comment]
While implementing this feature, I came across a problem which I was not aware of. Right now, the following happens when a PRG is opened (e.g., via the Open... menu or by providing a command line argument). 

1. A new emulator instance is created
2. The PRG file is converted to a disk and inserted into the drive
3. A new NSWindow is created

The problem is that 2. has to happen immediately after the emulator instance has been created and this is when the C64 has just begun to execute the Kernel Rom  start up code. This is not a problem when the PRG file gets converted to a disk, because the Kernal code does no harm to the disk. However, if I flash the PRG file into memory, the Kernal code will overwrite the flashed data in its memory initializing routine. At the moment, I only see very ugly solutions, i.e., the GUI would have to wait for a couple of seconds before it flashes the program. Such solutions are prone to errors, because they assume a certain execution speed of the other process. If the emulator thread get delayed for some reasons, the flashed file would still be overwritten. Considering the options, I suggest we better stick to the current solution ( convert PRGs and P00s to disks, if user dialogs are disabled).

**Comment by mlund** [comment]
I follow you and agree that an arbitrary guess for the kernal startup time is inelegant. Perhaps one could skip step 1 and have a menu option like "Execute PRG..." or drag-and-drop a PRG on to an _existing_ emulator instance? In that case the _user_ decides when to flash the PRG. This would also be useful in that the executed code can access and operate on the existing state of the emulator.

**Comment by dirkwhoffmann** [comment]
OK, I modified the drag and drop behavior. 

When mount dialogs are disabled and the user drags in a PRG or P00 file, it is directly flashed into memory. This was easy to implement, because drag and drop operates on a running emulator instance. 

The new code is already checked in (branch V2.3), but hasn't been tested thoroughly, yet.

**Comment by dirkwhoffmann** [comment]
BTW, how did you manage to see the debug output 

Disk: Encoded 61488 bits (7686 bytes) for tracks 1 to 17.  etc.
...

I'm asking, because in previous OS X versions, the debug output was dumped into the console and was easily viewable via the console app. In later macOS versions, Apple changed the logging system with the drawback that the messages don't appear in the console app any more (at least I didn't find them there). Did you run the emulator in XCode or did you use an older OS X?

**Comment by mlund** [comment]
Great, I'll check it out! Regarding the log, I just ran from a terminal:

~~~~bash
cd VirtualC64.app/Contents/MacOS/
./VirtualC64 ...
~~~~

**Comment by mlund** [comment]
I just tested 2.3beta using Xcode and Drag-and-drop is very convenient, thanks. I had access only to a $0801 file and I see "RUN" being typed in automatically which makes sense. Would this work also for code loaded to arbitrary positions, say $C000 where I would normally call "SYS49152"?

**Comment by mlund** [comment]
Yes, [Turbo assembler](http://style64.org/file/Turbo_Macro_Pro_Sep06_c64-STYLE.zip) loads at a non $0801 location (SYS 8*4096). I have it only as a D64 file and hoping to export a PRG from an attached disk, I used "Save as..." in the disk menu, a debug assertion kicked in:

~~~~cpp
int
D64Archive::offset(int track, int sector)
{
    assert(1 <= track && track <= 42); # <---
    assert(sector < D64Map[track].numberOfSectors);
    
    return D64Map[track].offset + (sector * 256);
}
~~~~

If using "RUN" for $0801 I think it would be consistent to use "SYS ..." for arbitrary machine code (although the PRG format *could* contain only data, but that would probably be rare). Perhaps a tick-box could determine the behavior?

**Comment by dirkwhoffmann** [comment]
OK, I'll try to implement it that way and I will fix the crash. However, I would like to avoid a check-box in order to keep the emulator's user interface simple.

**Comment by mlund** [comment]
Yes, better keep it simple as this feature will not likely be used by many :-) Thanks for adding this!

**Comment by mlund** [comment]
Thinking a bit more, perhaps it's better to skip "RETURN" and let the user do that if needed. In this way, one could use the feature to load up the C64 with data files (graphics, music etc.) and selectively apply RUN/SYS for code. I could imagine dragging multiple files in, but again, these features are for C64 software development purposes of interest (maybe) only to a small number of users(?)

**Comment by dirkwhoffmann** [comment]
Test case: 
[turbo.prg.zip](https://github.com/dirkwhoffmann/virtualc64/files/2216836/turbo.prg.zip)

**Comment by mlund** [comment]
I just tested `turbo.prg` with c8b5d36. I like the implementation and the code readability 👍

**Comment by dirkwhoffmann** [comment]
Fixed in 2.3 beta 3 

http://www.dirkwhoffmann.de/virtualC64/VirtualC64_2.3_beta3.zip
